### PR TITLE
[FLINK-34234] Bumping version of flatten plugin to 1.2.7

### PR DESCRIPTION
**Pull request by snuyanzin**
~~As a test [1] there was first a commit with enabling `ShadeOptionalChecker` however without adding `optional`. As a result it detects the issue i a number of places.~~

UPD: Since bumping version of flatten plugin helps I reverted changes (to not force push) and added version bump itself to 1.2.7
also keep a couple of hotfix commits as a result of review

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -49,8 +50,18 @@ jobs:
              -Plicense-check \
             | tee ${{ env.MVN_BUILD_OUTPUT_FILE }}
 
+          mvn ${MVN_COMMON_OPTIONS} dependency:tree -Plicense-check | tee ${{ env.MVN_DEPENDENCY_PLUGIN_OUTPUT_FILE }}
+
       - name: Check licensing
         run: |
-          mvn ${MVN_COMMON_OPTIONS} exec:java@check-licensing -N \
```
Comment (snuyanzin) [inline .github/workflows/ci.yml:54]: align with naming in pom

File: flink-shaded-asm-9/pom.xml [hunk 2]
```diff
@@ -43,24 +43,28 @@ under the License.
             <groupId>org.ow2.asm</groupId>
             <artifactId>asm</artifactId>
             <version>${asm.version}</version>
+            <optional>${flink.markBundledAsOptional}</optional>
```
Comment (snuyanzin) [inline flink-shaded-asm-9/pom.xml:46]: This naming is used in error logs in cases there is an issue

File: .github/workflows/ci.yml (hunk 1 above)
Comment (XComp) [reply .github/workflows/ci.yml:54]: That might deserve it's own hotfix commit

File: pom.xml [hunk 3]
```diff
@@ -341,21 +341,41 @@ under the License.
...
                     <dependency>
                         <groupId>org.apache.flink</groupId>
                         <artifactId>flink-ci-tools</artifactId>
-                        <version>1.17-SNAPSHOT</version>
+                        <version>1.18-SNAPSHOT</version>
                     </dependency>
                 </dependencies>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
```
Comment (XComp) [inline pom.xml:376]: why is this needed?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -49,8 +50,18 @@ jobs:
...
+          mvn ${MVN_COMMON_OPTIONS} exec:java@check-license -N \
```
Comment (XComp) [inline .github/workflows/ci.yml:57]: nit: can we move it into its own step rather than having it included in the Build step? ...to have a clearer separation and easier check what went wrong if something went wrong

File: flink-shaded-zookeeper-parent/flink-shaded-zookeeper-38/pom.xml [hunk 5]
```diff
@@ -128,6 +131,7 @@ under the License.
                 <groupId>org.apache.zookeeper</groupId>
                 <artifactId>zookeeper</artifactId>
                 <version>${zookeeper.version}</version>
+                <optional>${flink.markBundledAsOptional}</optional>
             </dependency>
             <dependency>
                 <groupId>io.netty</groupId>
```
Comment (XComp) [inline flink-shaded-zookeeper-parent/flink-shaded-zookeeper-38/pom.xml:137]: we don't need that for BOMs?!

File: pom.xml [hunk 6]
```diff
@@ -65,6 +65,7 @@ under the License.
         <netty.version>4.1.100.Final</netty.version>
         <jackson.version>2.15.3</jackson.version>
         <guava.version>32.1.3-jre</guava.version>
+        <flink.markBundledAsOptional>true</flink.markBundledAsOptional>
```
Comment (XComp) [inline pom.xml:68]: doesn't it make sense to upgrade the shade plugin to `3.4.1` here as well instead of having it only being upgraded for the jackson dependencies? 

my understanding is (according to [your comment in FLINK-34148](https://issues.apache.org/jira/browse/FLINK-34148?focusedCommentId=17809796&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17809796)) that the shade plugin stopped editing the dependency tree in 3.4.1. 

Maven 3.3+ didn't allow editing the dependency tree anymore (based on the description in FLINK-28016). Shouldn't we have run into issues already earlier because we upgraded Maven in CI to 3.8.2 even though we were still using an older version of the shade plugin? It seems that I'm missing something here. :thinking:

**Review by XComp — CHANGES_REQUESTED** [review]
I verified that https://github.com/apache/flink-shaded/commit/7caf8cd0c7ebe0a3167fa6a7e75697f6487627e6 which the [failed build](https://github.com/apache/flink-shaded/commit/7caf8cd0c7ebe0a3167fa6a7e75697f6487627e6) is based on matches the changes of this PR's first commit https://github.com/apache/flink-shaded/commit/ab87bb4ba53372b0ffc3275677842879a92af07b

File: pom.xml (hunk 3 above)
Comment (snuyanzin) [reply pom.xml:376]: that's the requirement to make `ShadeOptionalChecker` working
previous versions of `maven-dependency-plugin` do not mark dependencies `optional` in output of `mvn dependency:tree`

File: pom.xml (hunk 6 above)
Comment (snuyanzin) [reply pom.xml:68]: If I understand things around [MNG-5899](https://issues.apache.org/jira/browse/MNG-5899) correctly in maven they stopped reuse dependency reduced pom produced by shade plugin however they didn't forbid manipulation on api level.
At the same time on api level it is still possible to get some info e.g. list of dependencies from maven and then make some manipulation in code with this list of dependencies, this is how it was in maven-shade-plugin [1]. After [MSHADE-413](https://issues.apache.org/jira/browse/MSHADE-413) instead of manipulation on collection of objects received from maven they started to clone it first and then perform same manipulation on cloned version

[1] https://github.com/apache/maven-shade-plugin/pull/124/files#diff-083ef87367e79a2ea6b333846149588f4ce284ed930db1d87712fdbf9111e381L1066-L1112

File: .github/workflows/ci.yml (hunk 1 above)
Comment (snuyanzin) [reply .github/workflows/ci.yml:54]: yep, i was thinking about that however had some doubts... 
Since you've also mentioned this I think we should do that

Comment (snuyanzin) [reply .github/workflows/ci.yml:54]: extracted into a separate commit

File: .github/workflows/ci.yml (hunk 4 above)
Comment (snuyanzin) [reply .github/workflows/ci.yml:57]: IIUC it is already in a separate step...
Or do you mean a separate job?

**Comment by snuyanzin** [comment]
rebased since have to extract license related update into a separate commit

File: .github/workflows/ci.yml (hunk 4 above)
Comment (snuyanzin) [reply .github/workflows/ci.yml:55]: moved `mvn dependency:tree` to check shade optional step

File: flink-shaded-zookeeper-parent/flink-shaded-zookeeper-38/pom.xml (hunk 5 above)
Comment (snuyanzin) [reply flink-shaded-zookeeper-parent/flink-shaded-zookeeper-38/pom.xml]: That's not about BOMs
in general it does not make sense to add it for depedencies inside dependencyManagement
I removed it for others in dependencyManagement, thanks

File: pom.xml [hunk 7]
```diff
@@ -341,21 +342,41 @@ under the License.
...
-                    <mainClass>org.apache.flink.tools.ci.licensecheck.LicenseChecker</mainClass>
                     <includePluginDependencies>true</includePluginDependencies>
                     <includeProjectDependencies>false</includeProjectDependencies>
                 </configuration>
                 <dependencies>
                     <dependency>
                         <groupId>org.apache.flink</groupId>
                         <artifactId>flink-ci-tools</artifactId>
-                        <version>1.17-SNAPSHOT</version>
+                        <version>1.18-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (XComp) [reply pom.xml:369]: I'm wondering whether we could go for the released version as well to have a clearer reproducibility in case of errors.

[the remark this answers is not in the document]
Comment (snuyanzin) [reply pom.xml:369]: I think so.
I guess it could also go with its own hotfix commit

File: flink-shaded-zookeeper-parent/flink-shaded-zookeeper-38/pom.xml (hunk 5 above)
Comment (XComp) [reply flink-shaded-zookeeper-parent/flink-shaded-zookeeper-38/pom.xml]: Can you elaborate a bit here? It doesn't make sense because dependencies from the `<dependencyManagement/>` section would need to be explicitly added to a module pom to make the dependency available and only then would we have to add the `<optional/>` tag?!

Looks like we have 4 locations in `flink-dist/pom.xml` and `flink-filesystems/flink-s3-fs-presto/pom.xml` where this also applies. :thinking:

File: pom.xml (hunk 6 above)
Comment (XComp) [reply pom.xml:68]: I see, those were different issue. But that still leaves the question open whether we want to enable the more recent shade plugin version for all modules. :thinking:

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -51,6 +52,14 @@ jobs:
...
             -Dexec.args="${{ env.MVN_BUILD_OUTPUT_FILE }} $(pwd) ${{ env.MVN_VALIDATION_DIR }}" \
-            -Dlog4j.configurationFile=file://$(pwd)/tools/ci/log4j.properties
\ No newline at end of file
+            -Dlog4j.configurationFile=file://$(pwd)/tools/ci/log4j.properties
+
+      - name: Check shade optional dependencies
+        run: |
+            mvn ${MVN_COMMON_OPTIONS} dependency:tree -Plicense-check | tee ${{ env.MVN_DEPENDENCY_PLUGIN_OUTPUT_FILE }}
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply .github/workflows/ci.yml:61]: In fact it is...
The reason is that version suffix of several modules like guava, zookeeper is added only with this profile e.g. [1] and we need to know module name together with version suffix while parsing output of `dependency:tree`

https://github.com/apache/flink-shaded/blob/b4f9ed2dc85b3dcbb4bfc3bdb6866505bfae6893/flink-shaded-guava-32/pom.xml#L52-L59

[the remark this answers is not in the document]
Comment (snuyanzin) [reply .github/workflows/ci.yml:61]: may be we need to rename the profile itself... :thinking:

[the remark this answers is not in the document]
Comment (XComp) [reply .github/workflows/ci.yml:61]:
```
<!-- The license check and the optional shade check require the artifactId to match the directory that the module resides in.
```
hm, ok, if that's the case we could update [the comment](https://github.com/apache/flink-shaded/blob/392b737466b4be424873b48c0f31a999b8a34e70/pom.xml#L69) in the parent pom file to also mention the shade checker

[the remark this answers is not in the document]
Comment (XComp) [reply .github/workflows/ci.yml:61]: btw. the `exec-maven-plugin` allows to set System properties (see [docs](https://www.mojohaus.org/exec-maven-plugin/usage.html#pom-configuration-1)) which could be used to enable profiles implicitly (see [docs](https://maven.apache.org/guides/introduction/introduction-to-profiles.html#property)). :innocent:

File: flink-shaded-zookeeper-parent/flink-shaded-zookeeper-38/pom.xml (hunk 5 above)
Comment (snuyanzin) [reply flink-shaded-zookeeper-parent/flink-shaded-zookeeper-38/pom.xml]: May be I was not clear enough, sorry
By saying 
>it does not make sense to add it for depedencies inside dependencyManagement

I ment not in general.
Here there is already existing dependency on zookeeper for every zookeeper module in dependencies which overrides the one from dependency management (optional tag).
Netty here comes only as a dependency from zookeeper and once zookeeper is marked  optional maven started to think that netty is also becoming optional
I don't think this is the case for poms you've mentioned

File: pom.xml (hunk 6 above)
Comment (snuyanzin) [reply pom.xml:68]: that was the plan 
the only thing stopped me from making it here is that I was thinking about making ShadeOptionalChecker working which then could be backported to 18.x and then as a separate task sync maven-shade-plugin for all the modules

Or do you think it worth making it here?

Comment (XComp) [reply pom.xml:68]: fair point :+1: a follow-up Jira task is fine as well.

Comment (snuyanzin) [reply pom.xml:68]: yep, sure FLINK-34269

**Review by zentol — COMMENTED** [review]
Please see my comments in the ticket before continuing with this PR.

**Comment by snuyanzin** [comment]
Thanks for your feedback @zentol 
I checked locally with 1.2.7 flatten plugin and confirm that it resolves the issue

So I bumped the issue and currently reverted previous changes regarding shading...

May be one question which is still not clear to me (not directly related to perf regression)
Could you please clarify:
In confluence's creating flink-shaded release page it is mentioned that we need to use maven 3.2.5 to release.
Playing locally I was not able to find a difference between the result of usage 3.2.5 and 3.8.6.
Do we need still  apply ShadeOptionalChecker's approach like in Flink main repo to safely use 3.8.6 during release?

**Comment by zentol** [comment]
>  was not able to find a difference

That's because flink-shaded modules are all self-contained; no flink-shaded module depends on another flink-shaded module, due to which _right now_ it's not affected by the problem we had in the main repo.

This means you can probably use newer maven versions, but still shouldn't until we have added safeguards that either no interdependencies between modules exist or everything is marked as optional.
But I'd treat that as a separate issue altogether.

**Review by XComp — APPROVED** [review]
LGTM 👍 I verified that 1.2.7 is the newest 1.2.x patch version
